### PR TITLE
Integrate `nomad-substrate` into agents

### PR DESCRIPTION
**Pull request by luketchang**
## Motivation

Integrate `nomad-substrate` into rest of agents codebase. 

### High Level
- Breaks `nomad_core::Signers` into `nomad_ethereum::EthereumSigners` and `nomad_substrate::SubstrateSigners`
- Adds `nomad_substrate` crate which includes implemented `Home`, unimplemented `Replica` and `ConnectionManager`
- Moves `nomad_core::ChainCommunicationError` to nomad-base and makes it a wrapper across chain-specific error types defined in `chains/{X}`
- Removes `configuration/configs/multiVmTest.json` and unifies multi vm configuration tests by adding Avail to `configuration/configs/test.json`
- Adds new `substrate::TxSubmitterConf` to `configuration` crate (only supports local, gelato not for substrate)
- Adds substrate variant to `nomad-base::{Home, HomeIndexer}` and allows agent to configure substrate home objects from configuration and settings

### Details
- `nomad-substrate`
  - Adds implemented `Home` + unimplemented `Replica` and `ConnectionManager`
  - Add `SubstrateSigners` enum which supports local `PairSigner` (no AWS support yet)
  - Add macros for building `SubstrateHome` and `SubstrateHomeIndexer` from `nomad-base` provided configuration info
- `nomad-core`
  - Moves `nomad_core::ChainCommunicationError` to nomad-base and makes it a wrapper across chain-specific error types defined in `chains/{X}`, chains return their own error type which is wrapped by `nomad_base::ChainCommunicationError`
  - Moves `Signers` to `nomad-ethereum`, now being called `EthereumSigners` (no longer a unified `Signers` enum)
  - `nomad_core/signers.rs` only contains chain-agnostic signer traits (TODO: move SignerExt to base?)
- `nomad-base`
  - Re-exports `EthereumSigners` as new type `AttestationSigner` since all attestation signing is eth prefixed
  - Adds `Home` and `HomeIndexer` support for Substrate variant
  - Supports configuration of Substrate home objects from configuration and settings
- `configuration`
  - Add substrate chain to `configs/test.json` which has outbound connections to other chains but no inbound connections (no substrate replica support)

## PR Checklist

- [x] Added Tests
- [x] Updated Documentation
- [ ] Updated CHANGELOG.md for the appropriate package
- [ ] Ran PR in local/dev/staging

### CODE REVIEW

File: configuration/src/lib.rs [hunk 1]
```diff
@@ -135,44 +136,55 @@ impl NomadConfig {
...
+                        replica,
+                        name,
+                    );
+                    // Check for replicas found, but not configured
+                    eyre::ensure!(
+                        self.protocol.networks.get(replica).unwrap().connections.contains(name),
+                        "Replica named '{}' on core named '{}' not present in core's configured connections",
+                        replica,
+                        name
+                    );
```
Comment (luketchang) [inline configuration/src/lib.rs:186]: Compare L181-186 to previous statement (which I deleted) below. Believe this was a bug fix in our semantics. `connections` means **outbound connections** but previous code treated this as _inbound_ connections. Want second eye @lattejed @prestwich 

Before:
```
eyre::ensure!(
   self.protocol.networks.get(name).unwrap().connections.contains(replica),
   "Replica named '{}' on core named '{}' not present in core's configured connections",
   replica,
   name
);
```

After:
```
// Check for replicas found, but not configured
eyre::ensure!(
    self.protocol.networks.get(replica).unwrap().connections.contains(name),
    "Replica named '{}' on core named '{}' not present in core's configured connections",
    replica,
    name
);
```

Comment (lattejed) [reply configuration/src/lib.rs:186]: I'd say the change to `self.networks.contains(replica)` at L175 is correct. Previously it was just repeatedly re-checking the presence of `name` 

For your change at L182, I'm pretty sure the original was correct. We're just checking if `network` of `name` has `replica` in its list of `connections`. 

However I think `// Check for replicas found, but not configured` is unclear 

And I'm not sure if `// Check each core contains replicas ONLY for its listed connections` maps to the actual checks we're doing in that section of the validation.

Comment (lattejed) [reply configuration/src/lib.rs:186]: Actually, I think you're correct on the second part. Reasoning:

We're matching each core.replica.<network> to a protocol.<network>.connection and checking for the presence of name. Your inbound vs outbound correction holds imo

File: configuration/src/chains/ethereum/submitter/mod.rs [hunk 2]
```diff
@@ -39,9 +39,10 @@ pub enum TxSubmitterConf {
 
 impl From<crate::TxSubmitterConf> for TxSubmitterConf {
```
Comment (lattejed) [inline configuration/src/chains/ethereum/submitter/mod.rs:40]: I think it would be better to import this with an alias vs qualifying it with `crate::` -- the name overlap is pretty confusing at first glance

File: configuration/src/chains/substrate/mod.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+/// Local or relay-based transaction submission
+#[derive(Debug, Clone, PartialEq, serde::Deserialize)]
+#[serde(tag = "submitterType", content = "submitter", rename_all = "camelCase")]
+pub enum TxSubmitterConf {
+    /// Signer configuration for local signer
+    Local(SignerConf),
+}
+
+impl From<crate::TxSubmitterConf> for TxSubmitterConf {
```
Comment (lattejed) [inline configuration/src/chains/substrate/mod.rs:32]: Same comment as the one I made for `configuration/src/chains/ethereum/submitter/mod.rs`

File: configuration/src/core.rs [hunk 4]
```diff
@@ -27,40 +27,62 @@ pub struct EvmCoreContracts {
...
+    Substrate(SubstrateCoreDeployInfo),
 }
 
-impl CoreContracts {
+impl CoreDeploymentInfo {
     /// Get an iterator over the replicas present in this deploy
     pub fn replicas(&self) -> impl Iterator<Item = &String> {
         match self {
-            CoreContracts::Evm(contracts) => contracts.replicas.keys(),
+            CoreDeploymentInfo::Ethereum(contracts) => contracts.replicas.keys(),
```
Comment (lattejed) [inline configuration/src/core.rs:54]: Nit: Consistent use of brackets for these

File: configuration/src/lib.rs [hunk 5]
```diff
@@ -1,16 +1,17 @@
-//! Nomad Configuration crate with wasm bindings
+//! Nomad Configuration crate with wa&sm bindings
```
Comment (lattejed) [inline configuration/src/lib.rs:1]: Nit: Typo

File: configuration/src/secrets.rs [hunk 6]
```diff
@@ -160,15 +185,16 @@ mod test {
...
                 for config in secrets.tx_submitters.values() {
-                    assert_eq!(*config, default_config);
-                }
-                for config in secrets.rpcs.values() {
-                    assert!(matches!(*config, ChainConf::Ethereum { .. }));
+                    match config {
+                        TxSubmitterConf::Ethereum(conf) => {
+                            assert_eq!(*conf, default_config_ethereum)
+                        }
+                        TxSubmitterConf::Substrate(_) => continue,
```
Comment (lattejed) [inline configuration/src/secrets.rs:196]: Clarification: Why are we using a `continue` here

File: fixtures/test_secrets.json [hunk 7]
```diff
@@ -1,5 +1,5 @@
 {
-  "__COMMENT__": "matches configuration/configs/test.json",
+  "__COMMENT__": "matches configuration/configs/testMultiVm.json",
```
Comment (lattejed) [inline fixtures/test_secrets.json:2]: This file does not exist?

File: nomad-base/src/indexer.rs [hunk 8]
```diff
@@ -162,8 +162,10 @@ impl HomeIndexer for HomeIndexerVariants {
     async fn fetch_sorted_messages(&self, from: u32, to: u32) -> Result<Vec<RawCommittedMessage>> {
         match self {
             HomeIndexerVariants::Ethereum(indexer) => indexer.fetch_sorted_messages(from, to).await,
+            HomeIndexerVariants::Substrate(indexer) => {
```
Comment (lattejed) [inline nomad-base/src/indexer.rs:165]: Nit: Bracket use

File: nomad-base/src/settings/chains.rs [hunk 9]
```diff
@@ -137,16 +145,16 @@ impl ChainSetup {
...
                 Ok(HomeVariants::Ethereum(
-                    make_home(
-                        conf.clone(),
+                    nomad_ethereum::make_home(
+                        conn.clone(),
                         &ContractLocator {
                             name: self.name.clone(),
                             domain: self.domain,
-                            address: self.address,
+                            address: self.address.expect("must have address"),
```
Comment (lattejed) [inline nomad-base/src/settings/chains.rs:157]: Maybe flesh out this error more?

File: nomad-base/src/settings/chains.rs [hunk 10]
```diff
@@ -185,6 +208,7 @@ impl ChainSetup {
                 )
                 .into())
             }
+            ChainConf::Substrate(_) => unimplemented!("TODO"),
```
Comment (lattejed) [inline nomad-base/src/settings/chains.rs:211]: Are we leaving this intentionally for now?

File: nomad-base/src/settings/chains.rs [hunk 11]
```diff
@@ -197,20 +221,21 @@ impl ChainSetup {
...
-            ChainConf::Ethereum(conf) => Ok(ConnectionManagers::Ethereum(
+            ChainConf::Ethereum(conn) => Ok(ConnectionManagers::Ethereum(
                 make_conn_manager(
-                    conf.clone(),
+                    conn.clone(),
                     &ContractLocator {
                         name: self.name.clone(),
                         domain: self.domain,
-                        address: self.address,
+                        address: self.address.expect("must have address"),
```
Comment (lattejed) [inline nomad-base/src/settings/chains.rs:230]: cf L157

File: nomad-base/src/settings/mod.rs [hunk 12]
```diff
@@ -365,18 +364,22 @@ impl Settings {
...
         match &self.home.chain {
             ChainConf::Ethereum(conn) => Ok(HomeIndexerVariants::Ethereum(
-                make_home_indexer(
+                nomad_ethereum::make_home_indexer(
                     conn.clone(),
                     &ContractLocator {
                         name: self.home.name.clone(),
                         domain: self.home.domain,
-                        address: self.home.address,
+                        address: self.home.address.expect("must have address"),
```
Comment (lattejed) [inline nomad-base/src/settings/mod.rs:372]: cf nomad-base/src/settings/chains.rs L157

**Review by lattejed — COMMENTED** [review]
High level, open-ended question: 

Do we have enough / can we add more tests to this?

File: chains/nomad-ethereum/src/signer.rs [hunk 13]
```diff
@@ -115,28 +116,12 @@ impl Signer for Signers {
...
-        message: S,
-    ) -> Result<Signature, <Self as Signer>::Error> {
-        let mut signature = self.sign_message(message).await?;
-        signature.v = 28 - (signature.v % 2);
-        Ok(signature)
-    }
-}
-
-impl<T> SignerExt for T where T: Signer {}
-
```
[the remark this answers is not in the document]
Comment (lattejed) [reply chains/nomad-ethereum/src/signer.rs:139]: Found where this moved to later in the review

File: configuration/src/lib.rs (hunk 1 above)
Comment (luketchang) [reply configuration/src/lib.rs:186]: Cool so they are semantically correct but hard to parse. I will document these checks better, tbh I found them equally hard to parse when making the changes

File: configuration/src/secrets.rs (hunk 6 above)
Comment (luketchang) [reply configuration/src/secrets.rs:196]: Not really interested in checking substrate value as this is really a check to ensure ethereum default override worked

File: configuration/src/core.rs (hunk 4 above)
Comment (luketchang) [reply configuration/src/core.rs:54]: `cargo fmt` does this automatically. Would fail lint if these were not consistent

File: chains/nomad-substrate/src/lib.rs [hunk 14]
```diff
@@ -38,5 +41,72 @@ pub use utils::*;
...
+    conn: nomad_xyz_configuration::Connection,
+    name: &str,
+    timelag: Option<u8>,
+) -> color_eyre::Result<Box<dyn HomeIndexer>> {
+    let chain: SubstrateChains = name
+        .parse()
+        .unwrap_or_else(|_| panic!("Unrecognized chain name: {}", name));
+
+    match chain {
+        SubstrateChains::Avail => make_avail_home_indexer(conn, timelag).await,
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/lib.rs:110]: pretty sure this is `cargo fmt`

File: chains/nomad-substrate/src/signer.rs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    T: Config,
+    T::Signature: From<P::Signature>,
+    <T::Signature as Verify>::Signer: From<P::Public> + IdentifyAccount<AccountId = T::AccountId>,
+    P: Pair,
+{
+    async fn try_from_signer_conf(conf: &SignerConf) -> Result<Self> {
+        match conf {
+            SignerConf::HexKey(key) => {
+                let formatted_key = format!("0x{}", key.as_ref());
+                let pair = P::from_string(&formatted_key, None).unwrap(); // TODO: remove unwrap
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/signer.rs:34]: No I do not, will fix thanks

File: nomad-base/src/indexer.rs (hunk 8 above)
Comment (luketchang) [reply nomad-base/src/indexer.rs:165]: These are all decided by `fmt` not me :)

File: nomad-base/src/settings/chains.rs (hunk 10 above)
Comment (luketchang) [reply nomad-base/src/settings/chains.rs:211]: Yep

File: nomad-base/src/settings/chains.rs [hunk 16]
```diff
@@ -197,20 +221,21 @@ impl ChainSetup {
...
                     },
                     submitter_conf,
                     None, // Never need timelag for xapp connection manager
                     gas,
                 )
                 .await?,
             )),
+            ChainConf::Substrate(_) => unimplemented!("TODO"),
```
[the remark this answers is not in the document]
Comment (luketchang) [reply nomad-base/src/settings/chains.rs:238]: Yep same RE XCMs, not yet implemented. Will make unimplemented notes more clear ab that

File: nomad-core/src/traits/signer.rs [hunk 17]
```diff
@@ -0,0 +1,28 @@
+use async_trait::async_trait;
+use color_eyre::Result;
+use ethers::prelude::{Signature, Signer};
+use nomad_xyz_configuration::agent::SignerConf;
+
+/// Extension of ethers signer trait
+#[async_trait]
+pub trait SignerExt: Signer {
+    /// Sign message without eip 155
+    async fn sign_message_without_eip_155<S: Send + Sync + AsRef<[u8]>>(
```
[the remark this answers is not in the document]
Comment (luketchang) [reply nomad-core/src/traits/signer.rs:10]: No, this trait ext is for attestation signing attestations (which are not chain specific). EIP-155 is about binding sigs to chain ids but we need this trait for chain-agnostic attestations

[the remark this answers is not in the document]
Comment (luketchang) [reply nomad-core/src/traits/signer.rs:10]: This could probably live in `nomad-base` as that is where the `AttestationSigner` alias lives but not too big of deal

File: agents/watcher/src/watcher.rs [hunk 18]
```diff
@@ -560,9 +560,10 @@ impl NomadAgent for Watcher {
 
         let core = settings.as_ref().try_into_core("watcher").await?;
 
-        let signer =
-            Signers::try_from_signer_conf(&settings.base.attestation_signer.expect("signer"))
-                .await?;
+        let signer = AttestationSigner::try_from_signer_conf(
+            &settings.base.attestation_signer.expect("signer"),
```
[the remark this answers is not in the document]
Comment (luketchang) [reply agents/watcher/src/watcher.rs:564]: Fixed

File: configuration/src/bridge.rs [hunk 19]
```diff
@@ -28,10 +28,10 @@ pub struct DeployedCustomToken {
 /// EVM Bridge Contracts
 #[derive(Default, Debug, Clone, serde::Serialize, serde::Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct EvmBridgeContracts {
+pub struct EthereumBridgeDeploymentInfo {
     /// Contract Deploy Height
-    #[serde(default, deserialize_with = "deser_nomad_u64")]
-    pub deploy_height: u64,
+    #[serde(default, deserialize_with = "deser_nomad_u32")]
+    pub deploy_height: u32,
```
[the remark this answers is not in the document]
Comment (luketchang) [reply configuration/src/bridge.rs:34]: Because  `nomad-base::PageSettings` `from` (deploy height if fresh bootup/run) was `u32`. Doesn't matter which really just consistency

File: chains/nomad-substrate/src/signer.rs [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    async fn it_instantiates_and_signs() {
+        let conf = SignerConf::HexKey(
+            "1111111111111111111111111111111111111111111111111111111111111111"
+                .parse()
+                .unwrap(),
+        );
+        let signer = SubstrateSigners::<AvailConfig, ecdsa::Pair>::try_from_signer_conf(&conf)
+            .await
+            .unwrap();
+        let _sig = signer.sign("message".as_bytes());
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/signer.rs]: Yeah added line for `sig.verify(...)` to double check

File: chains/nomad-substrate/src/lib.rs [hunk 21]
```diff
@@ -38,5 +41,72 @@ pub use utils::*;
...
+pub(crate) enum SubstrateChains {
+    Avail,
+}
+
+impl FromStr for SubstrateChains {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_ref() {
+            "a" => Ok(Self::Avail),
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/lib.rs]: yep this was a mistake name was supposed to be "avail"

File: chains/nomad-substrate/bin/test.rs [hunk 22]
```diff
@@ -0,0 +1,29 @@
+use nomad_substrate::AvailConfig;
+use subxt::ext::scale_value::Value;
+use subxt::OnlineClient;
+
+#[tokio::main]
```
Comment (luketchang) [inline chains/nomad-substrate/bin/test.rs:5]: Ignore, will delete

File: chains/nomad-substrate/src/configs/avail.rs [hunk 23]
```diff
@@ -0,0 +1,439 @@
+#![allow(missing_docs)]
```
Comment (luketchang) [inline chains/nomad-substrate/src/configs/avail.rs:1]: Ignore this file, maintained by Avail team

**Review by lattejed — APPROVED** [review]
Doubling these here from our conversation:

Just want to double check that the substrate work doesn’t introduce *new* secret handling that needs to be kept out of logs

Also, If I run `cargo clippy --workspace --all-features -- -D warnings` locally I fail after a couple of errors. But the checks are passing. Can you double check this locally?



[10:51](https://nomad-xyz.slack.com/archives/D03E808T6ET/p1662565887286679)
Also, If I run cargo clippy --workspace --all-features -- -D warnings locally I fail after a couple of errors. But the checks are passing. Can you double check this locally?

File: chains/nomad-ethereum/CHANGELOG.md [hunk 24]
```diff
@@ -2,6 +2,8 @@
 
 ### Unreleased
 
+- Add `EthereumError` error enum to wrap ethers and gelato errors (ethereum-specific)
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-ethereum/CHANGELOG.md:5]: Yeah more meaning like ethereum rpcstyle specific

File: chains/nomad-ethereum/src/macros.rs [hunk 25]
```diff
@@ -42,19 +42,19 @@ macro_rules! report_tx {
         let dispatched = $provider
             .send_transaction($tx, None)
             .await
-            .map_err(|e| ChainCommunicationError::TxSubmissionError(e.into()))?;
+            .map_err(|e| crate::EthereumError::MiddlewareError(e.into()))?;
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-ethereum/src/macros.rs:45]: Yep this is correct. We submit local txs through using a set of middlewares so error sending tx would fall under that category. The returned error is wrapped in a bunch of generic types so we wrap it in box as MiddlewareError

File: chains/nomad-substrate/src/client.rs [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+            .block_hash(Some(block_number.into()))
+            .await?
+            .unwrap();
+
+        // Get updates from block
+        let update_events_res: Result<Vec<_>, _> = self
+            .events()
+            .at(Some(hash))
+            .await?
+            .find::<home::events::Update>() // TODO: remove dependency on avail metadata
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/client.rs:86]: Yes, we don't want this logic to be avail specific long-term but we're using the autogen structs given by avail here for convenience right now

File: chains/nomad-substrate/src/client.rs [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+            .into_iter()
+            .collect();
+
+        let update_events = update_events_res?;
+
+        // TODO: sort events
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/client.rs:92]: Yes we can leave for now, believe they are auto sorted. Sorting here would be quite involved anyways but made ticket as reminder

File: chains/nomad-substrate/src/home.rs [hunk 28]
```diff
@@ -0,0 +1,346 @@
...
+    <T as Config>::BlockNumber: TryInto<u32>,
+{
+    fn local_domain(&self) -> u32 {
+        self.domain
+    }
+
+    #[tracing::instrument(err, skip(self))]
+    async fn nonces(&self, destination: u32) -> Result<u32, <Self as Common>::Error> {
+        let nonce_address =
+            subxt::dynamic::storage("Home", "Nonces", vec![Value::u128(destination as u128)]);
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/home.rs:278]: We could but it's visually easier for me this way to namespace directly by what they are becaues these are names of pallets, items, and methods. Indirection would get confusing I think

File: chains/nomad-substrate/src/utils.rs [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+        if let subxt::Error::Runtime(subxt::error::DispatchError::Module(_)) = err {
+            return SubstrateError::TxNotExecuted(tx_in_block.extrinsic_hash().into());
+        }
+
+        SubstrateError::ProviderError(err)
+    })
+}
+
+/// Format signed update into scale value format
+pub fn format_signed_update_value(signed_update: &SignedUpdate) -> Value {
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/utils.rs:33]: The only way to tell correct formatting is really just if the chain accepts the tx. I can look into ways of testing locally or integration test but basically this was written as result of me hand massaging values into right shape based on error messages from the chain

[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/utils.rs:33]: Will think on this but not sure there's any good answers here

File: nomad-base/src/contract_sync/mod.rs [hunk 30]
```diff
@@ -165,6 +165,8 @@ where
 
                 let sorted_updates = indexer.fetch_sorted_updates(start, end).await?;
 
+                info!("Indexed block heights");
```
[the remark this answers is not in the document]
Comment (luketchang) [reply nomad-base/src/contract_sync/mod.rs:168]: I'll remove that my b

File: chains/nomad-substrate/src/macros.rs [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+            .await?;
+
+        info!(
+            method = $method,
+            tx_hash = ?pending_tx.extrinsic_hash(),
+            "Dispatched '{}' tx, waiting for inclusion.",
+            $method,
+        );
+
+        // TODO: can a tx deterministically revert here?
```
[the remark this answers is not in the document]
Comment (luketchang) [reply chains/nomad-substrate/src/macros.rs:17]: Yeah haven't seen it happen but will leave here for now in case we notice issues down the road
